### PR TITLE
/hack: improve shell script  in hack

### PR DESCRIPTION
**Pull request by tanjunchen**
/cc @justinsb  @mikesplain @rifelpet @robinpercy  
/hold

Let's talk about there changes 
i have run  ./hack/update-header.sh before i commited it.

### CODE REVIEW

File: hack/dev-build.sh [hunk 1]
```diff
@@ -110,9 +110,9 @@ echo ==========
 echo "Deleting cluster ${CLUSTER_NAME}. Elle est finie."
 
 kops delete cluster \
-  --name $CLUSTER_NAME \
```
Comment (tanjunchen) [inline hack/dev-build.sh]: Double quote to prevent globbing and word splitting

File: hack/make-apimachinery.sh [hunk 2]
```diff
@@ -18,27 +18,27 @@
...
-WORK_DIR=`mktemp -d`
+WORK_DIR=$(mktemp -d)
 
 cleanup() {
   chmod -R +w "${WORK_DIR}"
   rm -rf "${WORK_DIR}"
 }
 trap cleanup EXIT
 
-mkdir -p ${WORK_DIR}/go/
```
Comment (tanjunchen) [inline hack/make-apimachinery.sh]: Double quote to prevent globbing and word splitting

File: hack/make-gendocs.sh [hunk 3]
```diff
@@ -18,45 +18,45 @@
 
 . $(dirname "${BASH_SOURCE}")/common.sh
 
-WORK_DIR=`mktemp -d`
```
Comment (tanjunchen) [inline hack/make-gendocs.sh]: Use $(...) notation instead of legacy backticked `...`

File: hack/verify-spelling.sh [hunk 4]
```diff
@@ -12,21 +14,24 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-#!/bin/bash
```
Comment (tanjunchen) [inline hack/verify-spelling.sh]: The shebang must be on the first line

File: hack/pre-commit.sh [hunk 5]
```diff
@@ -12,5 +14,4 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-#!/bin/sh -eu
```
Comment (tanjunchen) [inline hack/pre-commit.sh]: The shebang must be on the first line

File: hack/publish-docs.sh [hunk 6]
```diff
@@ -14,7 +14,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-if ! [ -z $DEBUG ]; then
```
Comment (tanjunchen) [inline hack/publish-docs.sh]: recommend  [ -n .. ] instead of ! [ -z .. ]

File: hack/verify-bazel.sh [hunk 7]
```diff
@@ -14,7 +15,7 @@
 # limitations under the License.
 
 KOPS_ROOT=$(git rev-parse --show-toplevel)
-cd ${KOPS_ROOT}
```
Comment (tanjunchen) [inline hack/verify-bazel.sh]: in case cd fails.

File: hack/verify-gomod [hunk 8]
```diff
@@ -25,7 +25,7 @@ changes=$(git status --porcelain || true)
 if [ -n "${changes}" ]; then
   echo "ERROR: go modules are not up to date; please run: go mod tidy"
   echo "changed files:"
-  printf "${changes}\n"
```
Comment (tanjunchen) [inline hack/verify-gomod]: Don't use variables in the printf format string. Use printf "..%s.." "$foo"

File: hack/verify-apimachinery.sh [hunk 9]
```diff
@@ -23,7 +23,7 @@ export API_OPTIONS="--verify-only"
 if make apimachinery-codegen; then
 	echo "apimachinery is up to date"
 else
-	echo "\n FAIL: - the verify-apimachinery.sh test failed, apimachinery is not up to date"
-	echo "\n FAIL: - please run the command 'make apimachinery'"
+	echo -e "\n FAIL: - the verify-apimachinery.sh test failed, apimachinery is not up to date"
```
Comment (tanjunchen) [inline hack/verify-apimachinery.sh]: echo "\n FAIL: - the verify-apimachinery.sh test failed, apimachinery is not up to date"
```
\n FAIL: - the verify-apimachinery.sh test failed, apimachinery is not up to date
```
--------------->
echo -e "\n FAIL: - the verify-apimachinery.sh test failed, apimachinery is not up to date"
```

FAIL: - the verify-apimachinery.sh test failed, apimachinery is not up to date
```

File: hack/dev-build.sh (hunk 1 above)
Comment (tanjunchen) [reply hack/dev-build.sh]: nice

File: hack/go_install_from_commit.sh [hunk 10]
```diff
@@ -1,4 +1,5 @@
 #!/usr/bin/env bash
+
 # Copyright 2017 The Kubernetes Authors.
```
[the remark this answers is not in the document]
Comment (tanjunchen) [reply hack/go_install_from_commit.sh]: so we need to talk about it

[the remark this answers is not in the document]
Comment (joshbranham) [reply hack/go_install_from_commit.sh]: Yeah I am unsure so curious to hear from others here

**Comment by tanjunchen** [comment]
/cc @geojaz 
/cc @joshbranham  @justinsb  @rifelpet 
The so many scripts are changed before , I think it is not a good way.
we can open an new pr to improve the other files in the futrue , it is good for everybody.
/hold cancel

**Comment by justinsb** [comment]
These changes all look good, and we can always do more cleanups in other PRs.

We don't have a position that we know of on updating copyright years.

/approve
/lgtm

**Pull request merged**